### PR TITLE
JIT: fix assertion prop issues with subtypes

### PR DESCRIPTION
**Pull request by AndyAyersMS**
* don't use late args to destructure helper call
* create proper subtype constant kind when jitting

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib PTAL

The fix to the constant kind addresses part of #14471, where an exact type test guards a type cast. The jit can now eliminate the redundant helper call in the type cast during assertion prop, but not the redundant exact type test.

I have fixes for eliminating the redundant type test in the works, but they are a bit more elaborate, as the way assertion prop optimizes branches is kind of clunky.

Diffs show a few hits here and there:
```
Total bytes of diff: -147 (0.00% of base)
    diff is an improvement.

Top file improvements by size (bytes):
         -53 : System.Data.Common.dasm (0.00% of base)
         -30 : System.Private.CoreLib.dasm (0.00% of base)
         -29 : System.Private.DataContractSerialization.dasm (0.00% of base)
         -29 : System.Private.Xml.dasm (0.00% of base)
          -6 : System.Net.Security.dasm (0.00% of base)

5 total files with size differences (5 improved, 0 regressed), 124 unchanged.

Top method improvements by size (bytes):
         -37 (-18.32% of base) : System.Data.Common.dasm - System.Data.DefaultValueTypeConverter:ConvertFrom(ref,ref,ref):ref:this
         -29 (-7.14% of base) : System.Private.DataContractSerialization.dasm - System.Runtime.Serialization.SchemaExporter:ExportDataContract(ref):this
         -29 (-6.76% of base) : System.Private.Xml.dasm - System.Xml.XmlLoader:Load(ref,ref,bool):this
         -16 (-1.40% of base) : System.Data.Common.dasm - System.Data.Common.BigIntegerStorage:ConvertToBigInteger(ref,ref):struct
          -6 (-2.96% of base) : System.Net.Security.dasm - System.Net.Security.SecureChannel:MakeEx(ref):ref
          -5 (-0.48% of base) : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Byte,Int64][System.Byte,System.Int64]:.ctor(ref,ref):this (2 methods)
          -5 (-0.48% of base) : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Int16,Int64][System.Int16,System.Int64]:.ctor(ref,ref):this (2 methods)
          -5 (-0.48% of base) : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Int32,Int64][System.Int32,System.Int64]:.ctor(ref,ref):this (2 methods)
          -5 (-0.47% of base) : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Double,Int64][System.Double,System.Int64]:.ctor(ref,ref):this (2 methods)
          -5 (-0.44% of base) : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Vector`1,Int64][System.Numerics.Vector`1[System.Single],System.Int64]:.ctor(ref,ref):this (2 methods)
          -5 (-0.48% of base) : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Int64,Int64][System.Int64,System.Int64]:.ctor(ref,ref):this (2 methods)

Top method improvements by size (percentage):
         -37 (-18.32% of base) : System.Data.Common.dasm - System.Data.DefaultValueTypeConverter:ConvertFrom(ref,ref,ref):ref:this
         -29 (-7.14% of base) : System.Private.DataContractSerialization.dasm - System.Runtime.Serialization.SchemaExporter:ExportDataContract(ref):this
         -29 (-6.76% of base) : System.Private.Xml.dasm - System.Xml.XmlLoader:Load(ref,ref,bool):this
          -6 (-2.96% of base) : System.Net.Security.dasm - System.Net.Security.SecureChannel:MakeEx(ref):ref
         -16 (-1.40% of base) : System.Data.Common.dasm - System.Data.Common.BigIntegerStorage:ConvertToBigInteger(ref,ref):struct
          -5 (-0.48% of base) : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Int32,Int64][System.Int32,System.Int64]:.ctor(ref,ref):this (2 methods)
          -5 (-0.48% of base) : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Int64,Int64][System.Int64,System.Int64]:.ctor(ref,ref):this (2 methods)
          -5 (-0.48% of base) : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Byte,Int64][System.Byte,System.Int64]:.ctor(ref,ref):this (2 methods)
          -5 (-0.48% of base) : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Int16,Int64][System.Int16,System.Int64]:.ctor(ref,ref):this (2 methods)
          -5 (-0.47% of base) : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Double,Int64][System.Double,System.Int64]:.ctor(ref,ref):this (2 methods)
          -5 (-0.44% of base) : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Vector`1,Int64][System.Numerics.Vector`1[System.Single],System.Int64]:.ctor(ref,ref):this (2 methods)

11 total methods with size differences (11 improved, 0 regressed), 250188 unchanged.
```

**Comment by mikedn** [comment]
> don't use late args to destructure helper call

Thanks for cleaning up that!

File: src/jit/assertionprop.cpp [hunk 1]
```diff
@@ -1973,25 +1973,18 @@ AssertionInfo Compiler::optAssertionGenJtrue(GenTree* tree)
 
     GenTreeCall* call = op1->AsCall();
 
+    // See 3712, should we add chkcast...? also what about R2R
```
Comment (CarolEidt) [inline src/jit/assertionprop.cpp]: What is "3712" in reference to?

Comment (AndyAyersMS) [reply src/jit/assertionprop.cpp]: A left-behind reminder to myself to understand why the helper call lists at lines 1977 and 3712 differ, and why neither includes any R2R helpers.

Comment (AndyAyersMS) [reply src/jit/assertionprop.cpp]: Updated this comment to describe why the code is as it is.

**Pull request merged**